File: libgomp/testsuite/libgomp.c-c++-common/loop-1.c
extern
#ifdef __cplusplus
"C"
#endif
void abort (void);
#define N 256
int r;

void
foo (int *a)
{
  int i, j;
  #pragma omp loop bind(thread) order(concurrent) private (j) lastprivate (i) reduction(+:r) collapse(1)
  for (i = 0; i < N; i++)
    {
      j = i - 2;
      a[i] = j;
      r += j;
    }
}

void
bar (int *a)
{
  int i, j;
  #pragma omp loop bind(parallel) order(concurrent) private (j) lastprivate (i) reduction(+:r) collapse(1)
  for (i = 0; i < N; i++)
    {
      j = i;
      a[i] = j;
      r += j;
    }
}

void
baz (int *a)
{
  int i, j;
  #pragma omp loop bind(teams) order(concurrent) private (j) lastprivate (i) reduction(+:r)
  for (i = 0; i < N; i++)
    {
      j = i + 2;
      a[i] = j;
      r += j;
    }
}

int
main ()
{
  int a[N], i, j;
  foo (a);
  for (i = 0; i < N; ++i)
    if (a[i] != i - 2)
      abort ();
    else
      a[i] = -35;
  if (r != N * (N - 5) / 2)
    abort ();
  else
    r = 0;
  bar (a);
  for (i = 0; i < N; ++i)
    if (a[i] != i)
      abort ();
    else
      a[i] = -35;
  if (r != N * (N - 1) / 2)
    abort ();
  else
    r = 0;
  #pragma omp parallel loop private (j) lastprivate (i) reduction(+:r)
  for (i = 0; i < N; i++)
    {
      j = i + 4;
      a[i] = j;
      r += j;
    }
  if (i != N)
    abort ();
  for (i = 0; i < N; ++i)
    if (a[i] != i + 4)
      abort ();
    else
      a[i] = -35;
  if (r != N * (N + 7) / 2)
    abort ();
  else
    r = 0;
  #pragma omp parallel
  bar (a);
  for (i = 0; i < N; ++i)
    if (a[i] != i)
      abort ();
    else
      a[i] = -35;
  if (r != N * (N - 1) / 2)
    abort ();
  else
    r = 0;
  #pragma omp teams
  baz (a);
  for (i = 0; i < N; ++i)
    if (a[i] != i + 2)
      abort ();
    else
      a[i] = -35;
  if (r != N * (N + 3) / 2)
    abort ();
  else
    r = 0;
  #pragma omp teams loop order(concurrent) private (j) lastprivate (i) reduction(+:r) collapse(1)
  for (i = 0; i < N; i++)
    {
      j = i - 4;
      a[i] = j;
      r += j;
    }
  if (i != N)
    abort ();
  for (i = 0; i < N; ++i)
    if (a[i] != i - 4)
      abort ();
  if (r != N * (N - 9) / 2)
    abort ();
  return 0;
}

File: gcc/testsuite/gfortran.dg/gomp/loop-1.f90
! { dg-additional-options "-fdump-tree-original" }

implicit none
integer :: q, i, j
integer :: r
r = 0
!$omp loop bind(thread) reduction(default,+: r) collapse(2) order(concurrent), private(q) lastprivate(i)
do i = 1,4
do j = 1,4
  r = r + 1
  q = 5
end do
end do

!$omp teams loop bind(teams) collapse(2) order(concurrent), private(q) lastprivate(i) reduction(default,+: r) 
do i = 1,4
do j = 1,4
  r = r + 1
  q = 5
end do
end do

!$omp target teams loop bind(thread) reduction(+: r) collapse(2) order(concurrent), private(q) lastprivate(i)
do i = 1,4
do j = 1,4
  r = r + 1
  q = 5
end do
end do

!$omp parallel loop bind(thread) collapse(2) order(concurrent), private(q) lastprivate(i) reduction(default,+: r) 
do i = 1,4
do j = 1,4
  r = r + 1
  q = 5
end do
end do

!$omp target parallel loop bind(parallel) collapse(2) order(concurrent), private(q) lastprivate(i) reduction(default,+: r) 
do i = 1,4
do j = 1,4
  r = r + 1
  q = 5
end do
end do

end
 
! { dg-final { scan-tree-dump-times "#pragma omp target map\\(tofrom:i\\) map\\(tofrom:r\\)\[\r\n\]" 2 "original" } }
! { dg-final { scan-tree-dump-times "#pragma omp parallel shared\\(i\\) shared\\(r\\)\[\r\n\]" 2 "original" } }
! { dg-final { scan-tree-dump-times "#pragma omp teams shared\\(i\\) shared\\(r\\)\[\r\n\]" 2 "original" } }

! { dg-final { scan-tree-dump-times "#pragma omp loop private\\(q\\) lastprivate\\(i\\) reduction\\(\\+:r\\) order\\(concurrent\\) collapse\\(2\\) bind\\(parallel\\)" 1 "original" } }
! { dg-final { scan-tree-dump-times "#pragma omp loop private\\(q\\) lastprivate\\(i\\) reduction\\(\\+:r\\) order\\(concurrent\\) collapse\\(2\\) bind\\(teams\\)" 1 "original" } }
! { dg-final { scan-tree-dump-times "#pragma omp loop private\\(q\\) lastprivate\\(i\\) reduction\\(\\+:r\\) order\\(concurrent\\) collapse\\(2\\) bind\\(thread\\)" 3 "original" } }
